### PR TITLE
Java multiton

### PR DESCRIPTION
**Pull request by mcarcaso**
Mostly untested but it compiles and nanos starts happily so I don't think it's doing any harm. Just want to see if you're ok with the approach I'm taking with this.

### CODE REVIEW

**Comment by adamvy** [comment]
Seems reasonable.  I'm wondering if there's alternatives to using a Map for arguments, but I haven't come up with anything.

**Comment by mcarcaso** [comment]
Yeah I’m torn too. Maps are kind of annoying to work with in java but I
think they’re the most intuitive option for what I’m trying to accomplish
here. Im definitely open to changing this as we see how it unfolds but at
the moment I can’t think of anything that’s any better than a map.

**Comment by mcarcaso** [comment]
Renamed Factory to ContextFactory. I'll merge this in the morning if nobody has any objections.

File: src/foam/core/ClassInfoImpl.java [hunk 1]
```diff
@@ -61,10 +61,14 @@ public ClassInfo getParent() {
...
   }
 
-  public ClassInfo addProperty(PropertyInfo p) {
-    p.setClassInfo(this);
-    axioms.add(p);
-    axiomsByName_.put(p.getName(), p);
+  public ClassInfo addAxiom(Axiom a) {
+    // TODO: Should all axioms have setClassInfo? If not, create an interface
+    // that has setClassInfo and make PropertyInfo implement it.
+    if (a instanceof PropertyInfo) {
```
Comment (kgrgreer) [inline src/foam/core/ClassInfoImpl.java]: Follow FOAM style guideline for spacing.

File: src/foam/core/ContextFactory.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.core;
+
+public interface ContextFactory<T> {
```
Comment (kgrgreer) [inline src/foam/core/ContextFactory.java]: We also have XFactory: https://github.com/foam-framework/foam2/blob/f383a842c33d46772cdc81ddd13b89eb0490221a/src/foam/core/XFactory.java

Context is called X everywhere else.

File: src/foam/core/MultitonInfo.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  public T getInstance(Map<String, Object> args, X x) {
+    Object key = args.get(p.getName());
+    if ( ! instanceMap.containsKey(key) ) {
+      try {
+        Class<T> type = (Class<T>)p.getClassInfo().getObjClass();
+        T obj = type.newInstance();
+        ((ContextAware)obj).setX(x);
+        for (Map.Entry<String, Object> entry : args.entrySet()) {
```
Comment (kgrgreer) [inline src/foam/core/MultitonInfo.java]: spacing

File: src/foam/core/MultitonInfo.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  public MultitonInfo(String name, PropertyInfo p) {
+    this.name = name;
+    this.p = p;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public T getInstance(Map<String, Object> args, X x) {
```
Comment (kgrgreer) [inline src/foam/core/MultitonInfo.java]: Isn't thread safe so could cause duplicate creates.

File: src/foam/core/SimpleFacetManager.java [hunk 5]
```diff
@@ -15,17 +18,44 @@ public Object getInstanceOf(Object value, Class type, X x) {
...
+        ClassInfo classInfo = (ClassInfo)method.invoke(null);
+
+        Object f = null;
+        if ( x.get(classInfo.getId() + "_Factory") != null ) {
+          f = x.get(classInfo.getId() + "_Factory");
+        } else if ( classInfo.getAxiomsByClass(ContextFactory.class).size() == 1 ) {
+          f = classInfo.getAxiomsByClass(ContextFactory.class).get(0);
+        }
+
+        if (f != null) {
```
Comment (kgrgreer) [inline src/foam/core/SimpleFacetManager.java]: spacing

File: src/foam/core/SimpleFacetManager.java [hunk 6]
```diff
@@ -15,17 +18,44 @@ public Object getInstanceOf(Object value, Class type, X x) {
...
       if ( java.lang.reflect.Modifier.isAbstract(type.getModifiers()) ) {
         type = (Class<T>) Class.forName(type.getName() + "Impl");
       }
 
+      try {
+        java.lang.reflect.Method method = type.getMethod("getOwnClassInfo");
+        ClassInfo classInfo = (ClassInfo)method.invoke(null);
+
+        Object f = null;
+        if ( x.get(classInfo.getId() + "_Factory") != null ) {
```
Comment (kgrgreer) [inline src/foam/core/SimpleFacetManager.java]: Add comments to explain what's going on.

File: src/foam/core/ContextFactory.java (hunk 2 above)
Comment (mcarcaso) [reply src/foam/core/ContextFactory.java]: Yeah this is a little different than XFactory and I'm not sure what else to call this. Do you have any suggestions? XFObjectFactory?

File: src/foam/core/ClassInfoImpl.java (hunk 1 above)
Comment (mcarcaso) [reply src/foam/core/ClassInfoImpl.java]: Done

File: src/foam/core/MultitonInfo.java (hunk 3 above)
Comment (mcarcaso) [reply src/foam/core/MultitonInfo.java]: Done

File: src/foam/core/MultitonInfo.java (hunk 4 above)
Comment (mcarcaso) [reply src/foam/core/MultitonInfo.java]: Good eye. Made this method synchronized which should do the trick.

File: src/foam/core/SimpleFacetManager.java (hunk 5 above)
Comment (mcarcaso) [reply src/foam/core/SimpleFacetManager.java]: Done

File: src/foam/core/SimpleFacetManager.java (hunk 6 above)
Comment (mcarcaso) [reply src/foam/core/SimpleFacetManager.java]: Done. Let me know if it's still confusing.

**Pull request merged**